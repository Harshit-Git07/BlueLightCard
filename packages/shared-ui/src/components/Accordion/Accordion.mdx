import { Meta, Canvas, Controls } from '@storybook/blocks';

import * as Accordion from './Accordion.stories';

<Meta of={Accordion} />

# Accordion

- [Overview](#overview)
- [Props](#props)
- [Usage](#usage)

## Overview
Overview description of the Accordion component...

## Props
<Controls />

## Usage
➡️ Usage instruction one

➡️ Usage instruction two

➡️ Usage instruction three

➡️ Usage instruction four

## Variants
List of variants can be shown below here.

### Accordion Box
<Canvas of={Accordion.AccordionBox} />

### Accordion Box Markdown Bold
<Canvas of={Accordion.AccordionBoxMarkdownBold} />
