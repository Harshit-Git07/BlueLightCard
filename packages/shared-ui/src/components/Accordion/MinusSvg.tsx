import { FC } from 'react';

const MinusSvg: FC = () => (
  <svg width="14" height="14" viewBox="0 0 14 14" fill="none" xmlns="http://www.w3.org/2000/svg">
    <path
      d="M13.5 7C13.5 7.48398 13.0531 7.875 12.5 7.875H1.5C0.946875 7.875 0.5 7.48398 0.5 7C0.5 6.51602 0.946875 6.125 1.5 6.125H12.5C13.0531 6.125 13.5 6.51602 13.5 7Z"
      fill="currentColor"
    />
  </svg>
);

export default MinusSvg;
