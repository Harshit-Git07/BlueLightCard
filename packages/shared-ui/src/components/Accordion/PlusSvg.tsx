import { FC } from 'react';

const PlusSvg: FC = () => (
  <svg width="16" height="16" viewBox="0 0 16 16" fill="none" xmlns="http://www.w3.org/2000/svg">
    <path
      d="M8.99999 1.33325H7.66666V7.33325H1.66666V8.66659H7.66666V14.6666H8.99999V8.66659H15V7.33325H8.99999V1.33325Z"
      fill="#000033"
    />
  </svg>
);

export default PlusSvg;
