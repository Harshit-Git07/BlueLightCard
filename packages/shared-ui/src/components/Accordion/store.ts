import { atom } from 'jotai';

export const initialiseAccordionAtom = () => {
  return {} as {
    [key: string]: string | undefined;
  };
};

export const accordionAtom = atom(initialiseAccordionAtom());
export type AccordionAtom = ReturnType<typeof initialiseAccordionAtom>;
