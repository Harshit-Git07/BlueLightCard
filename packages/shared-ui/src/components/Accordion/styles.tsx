import { colours, fonts } from '../../tailwind/theme';
import { conditionalStrings } from '../../utils/conditionalStrings';

export const getClasses = (isOpen: boolean, isPartOfAGroup: boolean) => {
  const allStates = true;
  const isOpenAndGrouped = isOpen && isPartOfAGroup;
  const isClosedAndGrouped = !isOpen && isPartOfAGroup;

  const borderStyle = `border-b border-size-[1px] ${colours.borderOnSurfaceOutline}`;

  const accordionClasses = 'px-[16px]';

  const buttonClasses = conditionalStrings({
    [`flex justify-between items-center w-full py-[10px] text-left gap-[16px] ${colours.textOnSurface}`]:
      allStates,
    'mb-[16px]': isClosedAndGrouped,
    [borderStyle]: !isPartOfAGroup || isClosedAndGrouped,
  });

  const titleClasses = conditionalStrings({
    'whitespace-normal text-ellipses line-clamp-2': allStates,
    [fonts.bodySemiBold]: isPartOfAGroup,
    [fonts.body]: !isPartOfAGroup,
  });

  const contentWrapperClasses = conditionalStrings({
    'overflow-hidden transition-[height] duration-200 ease-in-out': allStates,
    [borderStyle]: isOpenAndGrouped,
  });

  const contentClasses = `pb-[24px] ${fonts.bodyLight} ${colours.textOnSurface}`;

  return { accordionClasses, buttonClasses, titleClasses, contentWrapperClasses, contentClasses };
};

export default getClasses;
