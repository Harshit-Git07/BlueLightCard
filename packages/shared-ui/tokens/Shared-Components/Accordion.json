{
  "accordion-divider-colour": {
    "light": {
      "value": "{colour.onSurface-outline.light}",
      "type": "color"
    },
    "dark": {
      "value": "{colour.onSurface-outline.dark}",
      "type": "color"
    }
  },
  "accordion-label-colour": {
    "light": {
      "value": "{colour.onSurface.light}",
      "type": "color"
    },
    "dark": {
      "value": "{colour.onSurface.dark}",
      "type": "color"
    }
  },
  "accordion-label-font": {
    "value": "{typography.body}",
    "type": "typography"
  },
  "accordion-text-font": {
    "value": "{typography.body-light}",
    "type": "typography"
  },
  "accordion-text-colour": {
    "light": {
      "value": "{colour.onSurface.light}",
      "type": "color"
    },
    "dark": {
      "value": "{colour.onSurface.dark}",
      "type": "color"
    }
  },
  "accordion-icon-colour": {
    "light": {
      "value": "{colour.onSurface.light}",
      "type": "color"
    },
    "dark": {
      "value": "{colour.onSurface.dark}",
      "type": "color"
    }
  }
}
