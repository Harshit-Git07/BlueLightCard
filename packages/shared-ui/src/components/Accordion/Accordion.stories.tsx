import React from 'react';
import { Meta, StoryFn } from '@storybook/react';
import Accordion from './';
import { Props } from './';
import Markdown from 'markdown-to-jsx';
// import IconListItem from '../IconListItem';

export default {
  title: 'Component System/Accordion',
  component: Accordion,
} as Meta;

const Default: StoryFn<Props> = (args) => (
  <Accordion {...args}>This is a simple accordion message.</Accordion>
);

const AccordionBoxMarkdownBoldTemplate: StoryFn<Props> = (args) => (
  <Accordion {...args}>
    <Markdown>
      {
        'These are **bold** terms and conditions for the above offer using *Markdown component* as children.'
      }
    </Markdown>
  </Accordion>
);

// const AccordionWithItemsTemplate: StoryFn<Props> = (args) => (
//   <Accordion {...args}>
//     <IconListItem
//       iconSrc="/assets/box-open-light-slash.svg"
//       title="Not valid on certain item(s)"
//       link="View details"
//       onClickLink={() => console.log('link clicked!')}
//     />
//     <IconListItem
//       iconSrc="/assets/circle-sterling-light.svg"
//       title="Only valid on full price items"
//     />
//   </Accordion>
// );

export const AccordionBox = Default.bind({});
AccordionBox.args = {
  title: 'Simple accordion message',
};

export const AccordionBoxMarkdownBold = AccordionBoxMarkdownBoldTemplate.bind({});
AccordionBoxMarkdownBold.args = {
  title: 'Terms and Conditions',
};

// export const AccordionBoxItemList = AccordionWithItemsTemplate.bind({});
// AccordionBoxItemList.args = {
//   title: 'Exclusions',
// };
