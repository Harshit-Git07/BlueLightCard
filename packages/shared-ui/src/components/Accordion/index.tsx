import { FC, useState } from 'react';
import MinusSvg from './MinusSvg';
import PlusSvg from './PlusSvg';

export type Props = {
  title: string;
  children: React.ReactNode;
};

const Accordion: FC<Props> = ({ title, children }) => {
  const [active, setActive] = useState(false);

  const handleToggle = () => {
    setActive(!active);
  };
  return (
    <div className="w-full leading-6">
      <button
        className="flex w-full text-left py-3 px-1 items-center border-b-[0.5px] border-b-[#DCDCDC] border-solid"
        onClick={() => handleToggle()}
      >
        <div className="w-full">
          <h4 className="font-['MuseoSans'] text-base font-medium text-[#1C1D22]">{title}</h4>
        </div>
        <div className="flex max-w-[40px] items-center justify-end">
          {active ? (
            <div style={{ width: 14, height: 14 }}>
              <MinusSvg />
            </div>
          ) : (
            <div style={{ width: 15, height: 15 }}>
              <PlusSvg />
            </div>
          )}
        </div>
      </button>

      <div
        className={`duration-200 ease-in-out ${
          active ? 'block' : 'hidden'
        } font-['MuseoSans'] py-3 text-base break-words font-light leading-5 whitespace-pre-wrap`}
      >
        {children}
      </div>
    </div>
  );
};

export default Accordion;
