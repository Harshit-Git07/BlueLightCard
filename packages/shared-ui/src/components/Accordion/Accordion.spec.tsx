import React, { act } from 'react';
import Accordion, { Props } from './';
import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom';
import userEvent from '@testing-library/user-event';
import { faEye, faEyeSlash } from '@fortawesome/pro-solid-svg-icons';

// Mocking the FontAwesome icons
jest.mock('@fortawesome/react-fontawesome', () => ({
  FontAwesomeIcon: ({ icon, className }: { icon: any; className?: string }) => (
    <span data-testid="font-awesome-icon" data-icon={icon.iconName} className={className}>
      FontAwesomeIcon
    </span>
  ),
}));

describe('Accordion Component', () => {
  const setup = (props: Partial<Props> = {}) => {
    const defaultProps: Props = {
      title: 'Test Accordion',
      children: <div>Accordion Content</div>,
      groupId: 'group-1',
      isOpenDefault: false,
    };

    return render(<Accordion {...defaultProps} {...props} />);
  };

  const isContentHidden = (element: Element | null) => {
    return element?.getAttribute('aria-hidden') === 'true';
  };

  it('renders the title and is closed by default', () => {
    setup();
    const contentElement = screen.getByText('Accordion Content').closest('[aria-hidden]');

    expect(screen.getByText('Test Accordion')).toBeInTheDocument();
    expect(isContentHidden(contentElement)).toBe(true);
    expect(screen.getByTestId('font-awesome-icon')).toHaveAttribute('data-icon', 'plus');
  });

  it('toggles open and closed when clicked', async () => {
    setup();
    const button = screen.getByRole('button', { name: 'Expand content' });
    const contentElement = screen.getByText('Accordion Content').closest('[aria-hidden]');

    // Initially closed
    expect(button).toHaveAttribute('aria-expanded', 'false');
    expect(isContentHidden(contentElement)).toBe(true);

    // Open accordion
    await act(async () => {
      await userEvent.click(button);
    });

    // After opening
    expect(button).toHaveAttribute('aria-expanded', 'true');
    expect(isContentHidden(contentElement)).toBe(false);

    // Click again to close
    await act(async () => {
      await userEvent.click(button);
    });

    expect(button).toHaveAttribute('aria-expanded', 'false');
    expect(isContentHidden(contentElement)).toBe(true);
  });

  it('applies correct aria attributes for accessibility', async () => {
    setup();

    const button = screen.getByRole('button', { name: 'Expand content' });
    const contentElement = screen.getByText('Accordion Content').closest('[aria-hidden]');

    expect(button).toHaveAttribute('aria-expanded', 'false');
    expect(contentElement).toHaveAttribute('aria-hidden', 'true');

    // Open accordion
    await act(async () => {
      await userEvent.click(button);
    });

    expect(button).toHaveAttribute('aria-expanded', 'true');
    expect(contentElement).toHaveAttribute('aria-hidden', 'false');
  });

  it('should disable focusable elements when closed and enable them when opened', async () => {
    setup({
      children: (
        <div>
          <button>Focusable Button</button>
          <a href="/">Focusable Link</a>
        </div>
      ),
    });

    const accordionButton = screen.getByRole('button', { name: 'Expand content' });
    const focusableButton = screen.getByText('Focusable Button');
    const focusableLink = screen.getByText('Focusable Link');

    // Initially, the focusable elements are hidden and disabled
    expect(focusableButton).toHaveAttribute('tabindex', '-1');
    expect(focusableLink).toHaveAttribute('tabindex', '-1');

    // Open the accordion
    await act(async () => {
      await userEvent.click(accordionButton);
    });

    // After opening, focusable elements should be enabled
    await waitFor(() => {
      expect(focusableButton).not.toHaveAttribute('tabindex');
      expect(focusableLink).not.toHaveAttribute('tabindex');
    });

    // Close the accordion
    await act(async () => {
      await userEvent.click(accordionButton);
    });

    // After closing, focusable elements should be disabled again
    await waitFor(() => {
      expect(focusableButton).toHaveAttribute('tabindex', '-1');
      expect(focusableLink).toHaveAttribute('tabindex', '-1');
    });
  });

  it('renders with custom icons if they are provided', async () => {
    setup({ openIcon: faEye, closeIcon: faEyeSlash });

    await act(async () => {
      await userEvent.click(screen.getByRole('button', { name: 'Expand content' }));
    });

    // Custom icons: faEye when closed, faEyeSlash when open
    expect(screen.getByTestId('font-awesome-icon')).toHaveAttribute('data-icon', 'eye-slash');
    await act(async () => {
      await userEvent.click(screen.getByRole('button', { name: 'Hide content' }));
    });
    expect(screen.getByTestId('font-awesome-icon')).toHaveAttribute('data-icon', 'eye');
  });

  it('supports initial default open state', () => {
    setup({ isOpenDefault: true });

    expect(screen.getByRole('button', { name: 'Hide content' })).toHaveAttribute(
      'aria-expanded',
      'true',
    );
    const accordionContent = screen.getByText('Accordion Content').closest('[aria-hidden]');
    expect(isContentHidden(accordionContent)).toBe(false);
  });

  it('should allow only one accordion to be opened at a time within the same group', async () => {
    render(
      <>
        <Accordion title="Accordion 1" groupId="group1">
          Content 1
        </Accordion>
        <Accordion title="Accordion 2" groupId="group1">
          Content 2
        </Accordion>
      </>,
    );

    const accordion1Button = screen.getAllByRole('button', { name: 'Expand content' })[0];
    const accordion2Button = screen.getAllByRole('button', { name: 'Expand content' })[1];

    const accordion1Content = screen.getByText('Content 1').closest('[aria-hidden]');
    const accordion2Content = screen.getByText('Content 2').closest('[aria-hidden]');

    // Open first accordion
    await act(async () => {
      await userEvent.click(accordion1Button);
    });
    expect(isContentHidden(accordion1Content)).toBe(false);
    expect(isContentHidden(accordion2Content)).toBe(true);

    // Open second accordion (first should close)
    await act(async () => {
      await userEvent.click(accordion2Button);
    });
    expect(isContentHidden(accordion1Content)).toBe(true);
    expect(isContentHidden(accordion2Content)).toBe(false);
  });

  it('should allow multiple accordions with different groupId to be opened at the same time', async () => {
    render(
      <>
        <Accordion title="Accordion 1" groupId="group1">
          Content 1
        </Accordion>
        <Accordion title="Accordion 2" groupId="group2">
          Content 2
        </Accordion>
      </>,
    );

    const accordion1Button = screen.getAllByRole('button', { name: 'Expand content' })[0];
    const accordion2Button = screen.getAllByRole('button', { name: 'Expand content' })[1];

    const accordion1Content = screen.getByText('Content 1').closest('[aria-hidden]');
    const accordion2Content = screen.getByText('Content 2').closest('[aria-hidden]');

    // Open first accordion
    await act(async () => {
      await userEvent.click(accordion1Button);
    });
    expect(isContentHidden(accordion1Content)).toBe(false);

    // Open second accordion (both should remain open)
    await act(async () => {
      await userEvent.click(accordion2Button);
    });
    expect(isContentHidden(accordion1Content)).toBe(false);
    expect(isContentHidden(accordion2Content)).toBe(false);
  });

  it('should allow all accordions with no groupId to be opened simultaneously', async () => {
    render(
      <>
        <Accordion title="Accordion 1">Content 1</Accordion>
        <Accordion title="Accordion 2">Content 2</Accordion>
      </>,
    );

    const accordion1Button = screen.getAllByRole('button', { name: 'Expand content' })[0];
    const accordion2Button = screen.getAllByRole('button', { name: 'Expand content' })[1];

    const accordion1Content = screen.getByText('Content 1').closest('[aria-hidden]');
    const accordion2Content = screen.getByText('Content 2').closest('[aria-hidden]');

    // Open first accordion
    await act(async () => {
      await userEvent.click(accordion1Button);
    });
    expect(isContentHidden(accordion1Content)).toBe(false);

    // Open second accordion (both should remain open)
    await act(async () => {
      await userEvent.click(accordion2Button);
    });
    expect(isContentHidden(accordion1Content)).toBe(false);
    expect(isContentHidden(accordion2Content)).toBe(false);
  });

  it('should adjust the content height when the accordion is open', async () => {
    const { container } = setup();

    const button = screen.getByRole('button', { name: 'Expand content' });

    fireEvent.click(button);

    const contentWrapper = container.querySelector('[aria-hidden="false"]');
    const content = contentWrapper?.firstElementChild;
    expect(contentWrapper).toBeInTheDocument();
    expect(content).toBeInTheDocument();

    Object.defineProperty(content, 'offsetHeight', {
      configurable: true,
      value: 100,
    });

    act(() => {
      window.dispatchEvent(new Event('resize'));
    });

    expect(contentWrapper).toHaveStyle('height: 100px');

    // Change the offsetHeight to simulate content change due to viewport resize
    Object.defineProperty(content, 'offsetHeight', {
      configurable: true,
      value: 200,
    });

    act(() => {
      window.dispatchEvent(new Event('resize'));
    });

    expect(contentWrapper).toHaveStyle('height: 200px');
  });

  it('should not adjust the content height when the accordion is closed', () => {
    const { container } = setup();

    const contentWrapper = container.querySelector('[aria-hidden="true"]');
    expect(contentWrapper).toBeInTheDocument();

    act(() => {
      window.dispatchEvent(new Event('resize'));
    });

    expect(contentWrapper).toHaveStyle('height: 0px');
  });

  it('should clean up resize event listener on unmount', () => {
    const removeEventListenerSpy = jest.spyOn(window, 'removeEventListener');
    const { unmount } = setup();

    unmount();

    expect(removeEventListenerSpy).toHaveBeenCalledWith('resize', expect.any(Function));
    removeEventListenerSpy.mockRestore();
  });
});
