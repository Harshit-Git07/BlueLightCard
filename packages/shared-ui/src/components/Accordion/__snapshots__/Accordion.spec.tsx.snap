// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion component should render an accordion with Markdown bold text 1`] = `
<div
  className="w-full leading-6"
>
  <button
    className="flex w-full text-left py-3 px-1 items-center border-b-[0.5px] border-b-[#DCDCDC] border-solid"
    onClick={[Function]}
  >
    <div
      className="w-full"
    >
      <h4
        className="font-museo text-base font-medium text-[#1C1D22]"
      >
        Terms and Conditions
      </h4>
    </div>
    <div
      className="flex max-w-[40px] items-center justify-end"
    >
      <div
        style={
          {
            "height": 15,
            "width": 15,
          }
        }
      >
        <svg
          fill="none"
          height="16"
          viewBox="0 0 16 16"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8.99999 1.33325H7.66666V7.33325H1.66666V8.66659H7.66666V14.6666H8.99999V8.66659H15V7.33325H8.99999V1.33325Z"
            fill="#000033"
          />
        </svg>
      </div>
    </div>
  </button>
  <div
    className="duration-200 ease-in-out hidden font-museo py-3 text-base break-words font-light leading-5 whitespace-pre-wrap text-black"
  >
    <span>
      This is an 
      <strong>
        accordion
      </strong>
       message.
    </span>
  </div>
</div>
`;

exports[`Accordion component should render an accordion with simple text 1`] = `
<div
  className="w-full leading-6"
>
  <button
    className="flex w-full text-left py-3 px-1 items-center border-b-[0.5px] border-b-[#DCDCDC] border-solid"
    onClick={[Function]}
  >
    <div
      className="w-full"
    >
      <h4
        className="font-museo text-base font-medium text-[#1C1D22]"
      >
        Simple accordion message
      </h4>
    </div>
    <div
      className="flex max-w-[40px] items-center justify-end"
    >
      <div
        style={
          {
            "height": 15,
            "width": 15,
          }
        }
      >
        <svg
          fill="none"
          height="16"
          viewBox="0 0 16 16"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8.99999 1.33325H7.66666V7.33325H1.66666V8.66659H7.66666V14.6666H8.99999V8.66659H15V7.33325H8.99999V1.33325Z"
            fill="#000033"
          />
        </svg>
      </div>
    </div>
  </button>
  <div
    className="duration-200 ease-in-out hidden font-museo py-3 text-base break-words font-light leading-5 whitespace-pre-wrap text-black"
  >
    This is a simple accordion message.
  </div>
</div>
`;
